{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mFailed to start the Kernel. \n",
      "\u001b[1;31mJupyter server crashed. Unable to connect. \n",
      "\u001b[1;31mError code from Jupyter: 1\n",
      "\u001b[1;31musage: jupyter.py [-h] [--version] [--config-dir] [--data-dir] [--runtime-dir]\n",
      "\u001b[1;31m                  [--paths] [--json] [--debug]\n",
      "\u001b[1;31m                  [subcommand]\n",
      "\u001b[1;31m\n",
      "\u001b[1;31mJupyter: Interactive Computing\n",
      "\u001b[1;31m\n",
      "\u001b[1;31mpositional arguments:\n",
      "\u001b[1;31m  subcommand     the subcommand to launch\n",
      "\u001b[1;31m\n",
      "\u001b[1;31moptional arguments:\n",
      "\u001b[1;31m  -h, --help     show this help message and exit\n",
      "\u001b[1;31m  --version      show the versions of core jupyter packages and exit\n",
      "\u001b[1;31m  --config-dir   show Jupyter config dir\n",
      "\u001b[1;31m  --data-dir     show Jupyter data dir\n",
      "\u001b[1;31m  --runtime-dir  show Jupyter runtime dir\n",
      "\u001b[1;31m  --paths        show all Jupyter paths. Add --json for machine-readable\n",
      "\u001b[1;31m                 format.\n",
      "\u001b[1;31m  --json         output paths as machine-readable json\n",
      "\u001b[1;31m  --debug        output debug information about paths\n",
      "\u001b[1;31m\n",
      "\u001b[1;31mAvailable subcommands:\n",
      "\u001b[1;31m\n",
      "\u001b[1;31mJupyter command `jupyter-notebook` not found. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "import csv\n",
    "import threading\n",
    "import time\n",
    "import pandas as pd\n",
    "\n",
    "main_keyword_list = {}\n",
    "\n",
    "def extract_keywords(protein_list: list) -> None:\n",
    "    with open(\"keywords.csv\", \"a\") as f:\n",
    "        for protein in protein_list:\n",
    "            # create words dict\n",
    "            words = {}\n",
    "\n",
    "            # extract request\n",
    "            url = \"https://rest.uniprot.org/uniprotkb/\" + protein + \".json\"\n",
    "            r = requests.get(url)\n",
    "            request = r.json()\n",
    "\n",
    "            for i, val in enumerate(request[\"keywords\"]):\n",
    "                words[request[\"keywords\"][i][\"name\"]] = 1\n",
    "\n",
    "            main_keyword_list[protein] = words\n",
    "\n",
    "def extract_proteins(path: str) -> list:\n",
    "    proteins = []\n",
    "\n",
    "    # read in text file\n",
    "    with open(path, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    # take the first value from text file\n",
    "    for line in lines:\n",
    "        proteins.append(line.split('|')[0])\n",
    "\n",
    "    # return list\n",
    "    return proteins\n",
    "\n",
    "def explore_json(protein: str) -> None:\n",
    "    url = \"https://rest.uniprot.org/uniprotkb/\" + protein + \".json\"\n",
    "    r = requests.get(url)\n",
    "    request = r.json()\n",
    "    for key in request.keys():\n",
    "        print(key)\n",
    "        print(request[key])\n",
    "        print()\n",
    "\n",
    "\n",
    "def main():\n",
    "    proteins = extract_proteins(\"data/protein_aliases.txt\")\n",
    "    res_path = \"results/functions.txt\"\n",
    "\n",
    "    extract_functions(proteins[:10])\n",
    "\n",
    "    df = pd.DataFrame(main_keyword_list)\n",
    "    df = df.transpose()\n",
    "    # print(df.head())\n",
    "    print(df.count())\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just get the most common keywords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import csv\n",
    "import threading\n",
    "import time\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "\n",
    "def extract_keywords(protein_list: list, res: list):\n",
    "    for protein in protein_list:\n",
    "        # create words dict\n",
    "\n",
    "        # extract request\n",
    "        url = \"https://rest.uniprot.org/uniprotkb/\" + protein + \".json\"\n",
    "        r = requests.get(url)\n",
    "        request = r.json()\n",
    "\n",
    "        for i, val in enumerate(request[\"keywords\"]):\n",
    "            res.append(val[\"name\"])\n",
    "\n",
    "def extract_proteins(path: str) -> list:\n",
    "    proteins = []\n",
    "\n",
    "    # read in text file\n",
    "    with open(path, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    # take the first value from text file\n",
    "    for line in lines:\n",
    "        proteins.append(line.split('|')[0])\n",
    "\n",
    "    # return list\n",
    "    return proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins = extract_proteins(\"data/protein_aliases.txt\")\n",
    "res_path = \"results/functions.txt\"\n",
    "\n",
    "res_1 = []\n",
    "res_2 = []\n",
    "res_3 = []\n",
    "res_4 = []\n",
    "res_5 = []\n",
    "res_6 = []\n",
    "\n",
    "t1 = threading.Thread(target=extract_keywords, args=(proteins[:5000], res_1, ))\n",
    "t2 = threading.Thread(target=extract_keywords, args=(proteins[5000:10000], res_2, ))\n",
    "t3 = threading.Thread(target=extract_keywords, args=(proteins[10000:15000], res_3, ))\n",
    "t4 = threading.Thread(target=extract_keywords, args=(proteins[15000:20000], res_4, ))\n",
    "t5 = threading.Thread(target=extract_keywords, args=(proteins[20000:25000], res_5, ))\n",
    "t6 = threading.Thread(target=extract_keywords, args=(proteins[25000:], res_6, ))\n",
    "\n",
    "\n",
    "t1.start()\n",
    "t2.start()\n",
    "t3.start()\n",
    "t4.start()\n",
    "t5.start()\n",
    "t6.start()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'Reference proteome': 30875, 'Proteomics identification': 13651, 'Phosphoprotein': 10329, 'Membrane': 8060, 'Cytoplasm': 7829, 'Acetylation': 6250, 'Alternative splicing': 5855, '3D-structure': 5608, 'Nucleus': 5566, 'Repeat': 5035, 'Metal-binding': 4721, 'Transmembrane': 4447, 'Transmembrane helix': 4348, 'Nucleotide-binding': 3951, 'Direct protein sequencing': 3733, 'Coiled coil': 3654, 'Signal': 3622, 'Transport': 3379, 'Mitochondrion': 2963, 'ATP-binding': 2953, 'Hydrolase': 2912, 'Glycoprotein': 2890, 'Transferase': 2832, 'Disulfide bond': 2825, 'Ubl conjugation': 2801, 'Cell membrane': 2397, 'Zinc': 2223, 'Isopeptide bond': 2125, 'Disease variant': 1797, 'Cytoskeleton': 1787, 'Endoplasmic reticulum': 1737, 'RNA-binding': 1710, 'Secreted': 1630, 'Methylation': 1573, 'Transit peptide': 1471, 'Protein transport': 1443, 'Transcription': 1394, 'Calcium': 1333, 'Oxidoreductase': 1315, 'Transcription regulation': 1291, 'Zinc-finger': 1204, 'Kinase': 1152, 'Ribonucleoprotein': 1144, 'Cell projection': 1120, 'Golgi apparatus': 1118, 'Lipid metabolism': 1100, 'Magnesium': 1085, 'DNA-binding': 1029, 'Lipoprotein': 988, 'Mitochondrion inner membrane': 862, 'Cytoplasmic vesicle': 854, 'Ribosomal protein': 853, 'GTP-binding': 846, 'Endosome': 830, 'Ubl conjugation pathway': 809, 'Actin-binding': 802, 'Protease': 755, 'mRNA processing': 743, 'Cell cycle': 704, 'Developmental protein': 629, 'Cell adhesion': 616, 'Apoptosis': 613, 'mRNA splicing': 609, 'WD repeat': 607, 'Cell junction': 589, 'Ion transport': 586, 'Immunity': 582, 'Differentiation': 580, 'Serine/threonine-protein kinase': 563, 'Chaperone': 543, 'Iron': 539, 'SH3 domain': 517, 'Receptor': 515, 'Chromosome': 512, 'Protein biosynthesis': 491, 'Lysosome': 490, 'Cell division': 480, 'Microtubule': 472, 'Activator': 468, 'Synapse': 467, 'Host-virus interaction': 453, 'Ligase': 450, 'NAD': 449, 'Repressor': 440, 'Innate immunity': 435, 'Isomerase': 409, 'DNA damage': 408, 'Extracellular matrix': 389, 'Lipid-binding': 376, 'Lyase': 371, 'DNA repair': 367, 'Palmitate': 364, 'Electron transport': 362, 'EGF-like domain': 360, 'Flavoprotein': 357, 'Immunoglobulin domain': 356, 'Intellectual disability': 344, 'Manganese': 342, 'NADP': 337, 'FAD': 334, 'Calmodulin-binding': 332, 'Mitochondrion outer membrane': 326, 'Spliceosome': 320, 'Motor protein': 319, 'Mitosis': 317, 'Lipid biosynthesis': 312, 'Fatty acid metabolism': 307, 'Neurogenesis': 302, 'Signal-anchor': 298, 'Helicase': 294, 'Chromatin regulator': 277, 'Myristate': 276, 'ANK repeat': 274, 'Prenylation': 272, 'ER-Golgi transport': 269, 'TPR repeat': 268, 'Hydroxylation': 262, 'Neurodegeneration': 259, 'Methyltransferase': 258, 'Leucine-rich repeat': 258, 'Autophagy': 255, 'Acyltransferase': 255, 'Thiol protease': 249, 'Respiratory chain': 247, 'Muscle protein': 244, 'Proteasome': 243, 'S-adenosyl-L-methionine': 240, 'GTPase activation': 239, 'Protein phosphatase': 234, 'Guanine-nucleotide releasing factor': 233, 'LIM domain': 232, 'Carbohydrate metabolism': 232, 'Lipid transport': 229, 'Endocytosis': 226, 'Translocation': 219, 'Translation regulation': 210, 'mRNA transport': 208, 'Metalloprotease': 206, 'Zymogen': 205, 'Translocase': 205, 'Wnt signaling pathway': 203, 'SH2 domain': 194, 'Glycosyltransferase': 193, 'Peroxisome': 193, 'Biological rhythms': 189, 'Transducer': 189, 'Iron-sulfur': 188, 'Myosin': 186, 'Initiation factor': 186, 'Serine protease': 185, 'Intermediate filament': 185, 'Cleavage on pair of basic residues': 182, 'Heme': 173, 'Aminoacyl-tRNA synthetase': 172, 'Primary mitochondrial disease': 172, 'ADP-ribosylation': 169, 'Lipid degradation': 166, 'Redox-active center': 165, 'Glycolysis': 160, 'Stress response': 156, 'Protease inhibitor': 150, 'Proto-oncogene': 147, 'Ion channel': 147, 'Angiogenesis': 146, 'Sarcoplasmic reticulum': 142, 'Nucleotidyltransferase': 142, 'Hydrogen ion transport': 141, 'DNA replication': 141, 'Copper': 140, 'Allosteric enzyme': 140, 'Nuclease': 140, 'Multifunctional enzyme': 138, 'Tyrosine-protein kinase': 138, 'Pyridoxal phosphate': 137, 'tRNA processing': 135, 'Chromosomal rearrangement': 134, 'rRNA processing': 134, 'Cilium': 133, 'Steroid metabolism': 132, 'Spermatogenesis': 132, 'Centromere': 132, 'Potassium': 131, 'Glycosidase': 131, 'Epilepsy': 130, 'Inflammatory response': 129, 'S-nitrosylation': 129, 'Heparin-binding': 127, 'Tricarboxylic acid cycle': 127, 'Lectin': 126, 'Rotamase': 126, 'Deafness': 125, 'Ribosome biogenesis': 125, 'Sterol metabolism': 121, 'Sushi': 119, 'Calcium transport': 119, 'Blood coagulation': 118, 'Hemostasis': 118, 'Alternative initiation': 118, 'Tumor suppressor': 117, 'Cholesterol metabolism': 112, 'Coated pit': 112, 'Nitration': 112, 'Antiviral defense': 111, 'Citrullination': 110, '4Fe-4S': 110, 'Fatty acid biosynthesis': 110, 'Tight junction': 109, 'Dioxygenase': 107, 'Collagen': 105, 'SH3-binding': 105, 'Proteoglycan': 103, 'Microsome': 102, 'Phospholipid metabolism': 102, 'Nuclear pore complex': 101, 'Kinetochore': 100, 'Basement membrane': 99, 'Lipid droplet': 97, 'Exocytosis': 97, 'Glycogen metabolism': 95, 'Integrin': 94, 'Phospholipid biosynthesis': 94, 'Neuropathy': 93, 'Adaptive immunity': 93, 'Dwarfism': 93, 'Amino-acid biosynthesis': 92, 'Serine protease inhibitor': 91, 'Sodium': 90, 'GPI-anchor': 89, 'Steroid biosynthesis': 89, 'Aminopeptidase': 88, 'Annexin': 87, 'Oxidation': 86, 'Cilium biogenesis/degradation': 85, 'Ubiquinone': 84, 'Keratin': 83, 'Complement pathway': 83, 'DNA recombination': 83, 'Cardiomyopathy': 82, 'Calcium/phospholipid-binding': 82, 'Actin capping': 81, 'Endonuclease': 81, 'Growth factor': 79, 'Peroxidase': 77, 'Sodium transport': 77, 'Potassium transport': 76, 'Pyrrolidone carboxylic acid': 74, 'Dynein': 74, 'Voltage-gated channel': 74, 'Synaptosome': 74, 'RNA-mediated gene silencing': 72, 'Nucleosome core': 72, 'Postsynaptic cell membrane': 69, 'Bromodomain': 66, 'One-carbon metabolism': 66, 'Sterol biosynthesis': 65, 'Nucleotide metabolism': 65, 'Aminotransferase': 64, 'Alternative promoter usage': 64, 'rRNA-binding': 63, 'Laminin EGF-like domain': 63, 'Sulfation': 62, 'Monooxygenase': 62, 'tRNA-binding': 62, 'Exonuclease': 61, 'Selenocysteine': 60, 'Nonsense-mediated mRNA decay': 60, 'Chromosome partition': 60, 'Cataract': 60, 'Elongation factor': 59, 'DNA-directed RNA polymerase': 59, 'ATP synthesis': 58, 'Antiport': 57, 'cAMP': 56, 'Calcium channel': 56, 'Cholesterol biosynthesis': 56, 'Acute phase': 56, '2Fe-2S': 56, 'Homeobox': 55, 'Purine biosynthesis': 55, 'Chloride': 55, 'Glucose metabolism': 54, 'HDL': 54, 'Thiamine pyrophosphate': 54, 'Symport': 54, 'Chemotaxis': 53, 'Gamma-carboxyglutamic acid': 52, 'Meiosis': 52, 'Serine esterase': 52, 'Flagellum': 51, 'Antioxidant': 51, 'Mitochondrion nucleoid': 51, 'CF(0)': 50, 'Cell shape': 50, 'Thick filament': 49, 'Cytokine': 49, 'Notch signaling pathway': 49, 'Carboxypeptidase': 48, 'Transmembrane beta strand': 47, 'Growth regulation': 46, 'Non-syndromic deafness': 46, 'Pyruvate': 45, 'Charcot-Marie-Tooth disease': 44, 'Antimicrobial': 43, 'Host cell receptor for virus entry': 42, 'Heparan sulfate': 42, 'FMN': 42, 'Myogenesis': 42, 'Protein phosphatase inhibitor': 41, 'Vitamin C': 41, 'Decarboxylase': 41, 'Thiol protease inhibitor': 40, 'CBS domain': 40, 'Amyloid': 40, 'Hereditary spastic paraplegia': 39, 'Cobalt': 38, 'Glutamine amidotransferase': 37, 'Ubiquinone biosynthesis': 37, 'Kringle': 37, 'Heme biosynthesis': 37, 'Unfolded protein response': 36, 'Complement alternate pathway': 35, 'cAMP-binding': 35, 'Retinitis pigmentosa': 34, 'Sphingolipid metabolism': 34, 'Exosome': 34, 'Biotin': 33, 'Nucleotide biosynthesis': 33, 'Gluconeogenesis': 33, 'Leukodystrophy': 33, 'Diabetes mellitus': 33, 'Schiff base': 32, 'Signalosome': 32, 'Porin': 32, 'Sensory transduction': 32, 'Porphyrin biosynthesis': 32, 'Autocatalytic cleavage': 31, 'Congenital disorder of glycosylation': 31, 'Osteogenesis': 31, 'Phagocytosis': 31, 'Antibiotic': 31, 'Growth factor binding': 31, 'Kelch repeat': 30, 'Virion': 29, 'Limb-girdle muscular dystrophy': 29, 'Mitogen': 29, 'Hormone': 29, 'Cytolysis': 29, 'Methionine biosynthesis': 28, 'Gastrulation': 28, 'Pyrimidine biosynthesis': 28, 'Membrane attack complex': 28, 'Amyotrophic lateral sclerosis': 28, 'Lipoyl': 27, 'Viral nucleoprotein': 27, 'Ciliopathy': 27, 'Prostaglandin metabolism': 27, 'Iron transport': 27, 'Chloride channel': 27, 'VLDL': 26, 'Sodium/potassium transport': 26, 'Amino-acid transport': 26, 'Hereditary hemolytic anemia': 26, 'Signal transduction inhibitor': 26, 'Signal recognition particle': 26, 'Parkinsonism': 26, 'CF(1)': 26, 'Obesity': 25, 'Ligand-gated ion channel': 25, 'Oxygen transport': 25, 'Vision': 25, 'Pyridine nucleotide biosynthesis': 25, 'Chylomicron': 24, 'Purine salvage': 24, 'Sialic acid': 24, 'Glycogen biosynthesis': 24, 'Dystonia': 24, 'Thioester bond': 24, 'Molybdenum': 24, 'Fibrinolysis': 23, 'Prostaglandin biosynthesis': 23, 'Ectodermal dysplasia': 23, 'cGMP': 23, 'Telomere': 23, 'Glutathione biosynthesis': 22, 'Sugar transport': 22, 'Vasoactive': 22, 'Autism spectrum disorder': 22, 'Selenium': 21, 'Glycogen storage disease': 21, 'Fertilization': 21, 'GPI-anchor biosynthesis': 20, 'Biomineralization': 20, 'Isoprene biosynthesis': 20, 'Molybdenum cofactor biosynthesis': 20, 'G-protein coupled receptor': 20, 'Protein kinase inhibitor': 20, 'Ichthyosis': 20, 'Glutathionylation': 20, 'Amyloidosis': 20, 'Phosphopantetheine': 19, 'Potassium channel': 19, 'Hearing': 19, 'Metalloenzyme inhibitor': 19, 'Spinocerebellar ataxia': 19, 'Lissencephaly': 19, 'Branched-chain amino acid catabolism': 18, 'Diamond-Blackfan anemia': 18, 'Threonine protease': 18, 'Hypotrichosis': 18, 'Neurotransmitter degradation': 18, 'DNA-directed DNA polymerase': 18, 'Prenyltransferase': 18, 'Urea cycle': 18, 'Leigh syndrome': 17, 'GMP biosynthesis': 17, 'Copper transport': 17, 'Metalloprotease inhibitor': 17, 'Gap junction': 17, 'Congenital muscular dystrophy': 17, 'Nickel': 17, 'Phenylalanine catabolism': 17, 'Tyrosine catabolism': 17, 'Parkinson disease': 17, 'Topoisomerase': 16, 'TPQ': 16, 'Branched-chain amino acid biosynthesis': 16, 'Palmoplantar keratoderma': 16, 'Osteogenesis imperfecta': 16, 'Lithium': 15, 'Catecholamine metabolism': 15, 'Zinc transport': 15, 'Transcription termination': 15, 'Sodium channel': 15, 'Leukotriene biosynthesis': 15, 'Serine protease homolog': 15, 'mRNA capping': 14, 'RNA editing': 14, 'Serine biosynthesis': 14, 'Blood group antigen': 14, 'DNA synthesis': 14, 'Alzheimer disease': 14, 'Proline biosynthesis': 14, 'Tetrahydrobiopterin biosynthesis': 14, 'Glyoxylate bypass': 14, 'Peroxisome biogenesis': 13, 'LDL': 13, 'Pharmaceutical': 13, 'Aspartyl protease': 13, 'Hypusine': 13, 'Coenzyme A biosynthesis': 13, 'Anion exchange': 13, 'Hyaluronic acid': 13, 'Phospholipase A2 inhibitor': 13, 'Aortic aneurysm': 12, 'Steroid-binding': 12, 'Pentose shunt': 12, 'Plasminogen activation': 12, 'Cobalamin': 12, 'cGMP biosynthesis': 12, 'Retinol-binding': 12, 'Ehlers-Danlos syndrome': 11, 'Necrosis': 11, 'Detoxification': 11, 'Epidermolysis bullosa': 11, 'Thrombophilia': 11, 'Autism': 11, 'Triplet repeat expansion': 11, 'Primary microcephaly': 11, 'cAMP biosynthesis': 11, 'Iron storage': 11, 'Protein synthesis inhibitor': 11, 'Cyclin': 11, 'Albinism': 11, 'Glycation': 11, 'Myofibrillar myopathy': 10, 'Fucose metabolism': 10, 'Dipeptidase': 10, 'Vasodilator': 10, 'Williams-Beuren syndrome': 10, 'Microphthalmia': 10, 'Oogenesis': 10, 'Collagen degradation': 10, 'Tissue remodeling': 10, 'Hermansky-Pudlak syndrome': 10, 'Age-related macular degeneration': 10, 'Mucopolysaccharidosis': 10, 'Vitamin A': 10, 'Long QT syndrome': 10, 'Purine metabolism': 9, 'Bardet-Biedl syndrome': 9, 'Tumor antigen': 9, 'Nemaline myopathy': 9, 'Hydrogen peroxide': 9, 'Schizophrenia': 9, 'cGMP-binding': 9, 'Proline metabolism': 9, 'Leber hereditary optic neuropathy': 9, 'Corneal dystrophy': 9, 'Quinone': 9, 'Craniosynostosis': 9, 'Vitamin D': 9, 'Galactose metabolism': 8, 'Amidation': 8, 'Aromatic hydrocarbons catabolism': 8, 'Progressive external ophthalmoplegia': 8, 'Erythrocyte maturation': 8, 'Folate-binding': 8, 'LTQ': 8, 'Hypogonadotropic hypogonadism': 8, 'Phosphate transport': 8, 'Atrial fibrillation': 8, 'Bait region': 8, 'Arginine biosynthesis': 8, 'Intrahepatic cholestasis': 8, 'Complement activation lectin pathway': 7, 'Glaucoma': 7, 'Congenital myasthenic syndrome': 7, 'Neuropeptide': 7, 'Neurotransmitter transport': 7, 'Mannose-binding': 7, 'Gluconate utilization': 7, 'Peroxisome biogenesis disorder': 7, 'Eye lens protein': 7, 'Digestion': 7, 'Prion': 7, 'Hemoglobin-binding': 7, 'Dystroglycanopathy': 7, 'Glycerol metabolism': 7, 'Elliptocytosis': 6, 'Joubert syndrome': 6, 'Oncogene': 6, 'Cushing syndrome': 6, 'Emery-Dreifuss muscular dystrophy': 6, 'Deoxyribonucleotide synthesis': 6, '3Fe-4S': 6, 'SCID': 6, 'Steroidogenesis': 6, 'Cone-rod dystrophy': 6, 'Atrial septal defect': 6, 'Leber congenital amaurosis': 6, 'Hemolytic uremic syndrome': 6, 'Cysteine biosynthesis': 5, 'Polyamine biosynthesis': 5, 'Neuronal ceroid lipofuscinosis': 5, 'Catecholamine biosynthesis': 5, 'Allergen': 5, 'DNA condensation': 5, 'Gangliosidosis': 5, 'Vasoconstrictor': 5, 'Atherosclerosis': 5, 'Thyroid hormone': 5, 'Kallmann syndrome': 5, 'Inflammasome': 5, 'Brugada syndrome': 5, 'Immunoglobulin': 5, 'Cobalt transport': 5, 'Hyperlipidemia': 5, 'Systemic lupus erythematosus': 5, 'Hereditary nonpolyposis colorectal cancer': 5, 'MHC I': 5, 'Histidine biosynthesis': 5, 'Arginine metabolism': 5, 'DNA excision': 5, 'Hypusine biosynthesis': 4, 'Aicardi-Goutieres syndrome': 4, 'B-cell activation': 4, 'Behavior': 4, 'Phospholipid degradation': 4, 'Mineral balance': 4, 'Carnitine biosynthesis': 4, 'Dyskeratosis congenita': 4, 'Nephronophthisis': 4, 'Zellweger syndrome': 4, 'Histidine metabolism': 4, 'Peptide transport': 4, 'Keratinization': 4, 'Glutaricaciduria': 4, 'Chronic granulomatous disease': 4, 'Alport syndrome': 4, 'Chondrogenesis': 4, 'Chromophore': 4, 'Neurotransmitter biosynthesis': 3, 'Pyridoxine biosynthesis': 3, 'Maple syrup urine disease': 3, 'Congenital hypothyroidism': 3, 'Xylose metabolism': 3, 'Target cell membrane': 3, 'Target membrane': 3, 'Olfaction': 3, 'IgE-binding protein': 3, 'Hirschsprung disease': 3, 'Ascorbate biosynthesis': 3, 'Congenital dyserythropoietic anemia': 3, 'Asparagine biosynthesis': 3, 'Premature ovarian failure': 3, 'Metachromatic leukodystrophy': 3, 'Congenital generalized lipodystrophy': 3, 'IgG-binding protein': 3, 'MELAS syndrome': 3, 'Inositol biosynthesis': 3, 'Virus entry into host cell': 3, 'Melanin biosynthesis': 3, 'Amelogenesis imperfecta': 3, 'Niemann-Pick disease': 3, 'Rhizomelic chondrodysplasia punctata': 3, 'Osteopetrosis': 3, 'Primary ciliary dyskinesia': 3, 'Holoprosencephaly': 3, 'Xeroderma pigmentosum': 3, 'Spermidine biosynthesis': 2, 'Heterotaxy': 2, 'Hemophilia': 2, 'Asthma': 2, 'Bromination': 2, 'Meckel syndrome': 2, 'Antiviral protein': 2, 'Gout': 2, 'Viral RNA replication': 2, 'Congenital stationary night blindness': 2, 'Sexual differentiation': 2, 'Bacteriolytic enzyme': 2, 'Stickler syndrome': 2, 'Pyropoikilocytosis': 2, 'Septation': 2, 'Gaucher disease': 2, 'Dejerine-Sottas syndrome': 2, 'Ammonia transport': 2, 'Congenital erythrocytosis': 2, 'Kartagener syndrome': 2, 'Mast cell degranulation': 2, 'Desmin-related myopathy': 2, 'Neurotransmitter': 2, 'von Willebrand disease': 2, 'Viral penetration into host nucleus': 2, 'Lactation': 2, 'Cockayne syndrome': 2, 'Usher syndrome': 2, 'Mucolipidosis': 2, 'Familial hemophagocytic lymphohistiocytosis': 2, 'Fanconi anemia': 1, 'Capsid protein': 1, 'Translational shunt': 1, 'Congenital adrenal hyperplasia': 1, 'Luminescence': 1, 'Photoprotein': 1, 'Cytoplasmic inwards viral transport': 1, 'Microtubular inwards viral transport': 1, 'Short QT syndrome': 1, 'Stargardt disease': 1, 'Diabetes insipidus': 1, 'Hypotensive agent': 1, 'Growth arrest': 1, 'Senior-Loken syndrome': 1, 'Bence-Jones protein': 1, 'Vacuole': 1, 'Bernard Soulier syndrome': 1, 'Pseudohermaphroditism': 1, 'RNA-directed RNA polymerase': 1, 'Inhibition of host innate immune response by virus': 1, 'Inhibition of host interferon signaling pathway by virus': 1, 'Inhibition of host STAT1 by virus': 1, 'Inhibition of host STAT2 by virus': 1, 'Viral immunoevasion': 1, 'Phenylketonuria': 1, 'Pregnancy': 1, 'Pontocerebellar hypoplasia': 1, 'Thioether bond': 1, 'Primary hypomagnesemia': 1, 'Bartter syndrome': 1})\n"
     ]
    }
   ],
   "source": [
    "all_words = res_1 + res_2 + res_3 + res_4 + res_5 + res_6\n",
    "c = Counter(all_words)\n",
    "print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Counter in module collections object:\n",
      "\n",
      "class Counter(builtins.dict)\n",
      " |  Counter(iterable=None, /, **kwds)\n",
      " |  \n",
      " |  Dict subclass for counting hashable items.  Sometimes called a bag\n",
      " |  or multiset.  Elements are stored as dictionary keys and their counts\n",
      " |  are stored as dictionary values.\n",
      " |  \n",
      " |  >>> c = Counter('abcdeabcdabcaba')  # count elements from a string\n",
      " |  \n",
      " |  >>> c.most_common(3)                # three most common elements\n",
      " |  [('a', 5), ('b', 4), ('c', 3)]\n",
      " |  >>> sorted(c)                       # list all unique elements\n",
      " |  ['a', 'b', 'c', 'd', 'e']\n",
      " |  >>> ''.join(sorted(c.elements()))   # list elements with repetitions\n",
      " |  'aaaaabbbbcccdde'\n",
      " |  >>> sum(c.values())                 # total of all counts\n",
      " |  15\n",
      " |  \n",
      " |  >>> c['a']                          # count of letter 'a'\n",
      " |  5\n",
      " |  >>> for elem in 'shazam':           # update counts from an iterable\n",
      " |  ...     c[elem] += 1                # by adding 1 to each element's count\n",
      " |  >>> c['a']                          # now there are seven 'a'\n",
      " |  7\n",
      " |  >>> del c['b']                      # remove all 'b'\n",
      " |  >>> c['b']                          # now there are zero 'b'\n",
      " |  0\n",
      " |  \n",
      " |  >>> d = Counter('simsalabim')       # make another counter\n",
      " |  >>> c.update(d)                     # add in the second counter\n",
      " |  >>> c['a']                          # now there are nine 'a'\n",
      " |  9\n",
      " |  \n",
      " |  >>> c.clear()                       # empty the counter\n",
      " |  >>> c\n",
      " |  Counter()\n",
      " |  \n",
      " |  Note:  If a count is set to zero or reduced to zero, it will remain\n",
      " |  in the counter until the entry is deleted or the counter is cleared:\n",
      " |  \n",
      " |  >>> c = Counter('aaabbc')\n",
      " |  >>> c['b'] -= 2                     # reduce the count of 'b' by two\n",
      " |  >>> c.most_common()                 # 'b' is still in, but its count is zero\n",
      " |  [('a', 3), ('c', 1), ('b', 0)]\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Counter\n",
      " |      builtins.dict\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add counts from two counters.\n",
      " |      \n",
      " |      >>> Counter('abbb') + Counter('bcc')\n",
      " |      Counter({'b': 4, 'c': 2, 'a': 1})\n",
      " |  \n",
      " |  __and__(self, other)\n",
      " |      Intersection is the minimum of corresponding counts.\n",
      " |      \n",
      " |      >>> Counter('abbb') & Counter('bcc')\n",
      " |      Counter({'b': 1})\n",
      " |  \n",
      " |  __delitem__(self, elem)\n",
      " |      Like dict.__delitem__() but does not raise KeyError for missing values.\n",
      " |  \n",
      " |  __iadd__(self, other)\n",
      " |      Inplace add from another counter, keeping only positive counts.\n",
      " |      \n",
      " |      >>> c = Counter('abbb')\n",
      " |      >>> c += Counter('bcc')\n",
      " |      >>> c\n",
      " |      Counter({'b': 4, 'c': 2, 'a': 1})\n",
      " |  \n",
      " |  __iand__(self, other)\n",
      " |      Inplace intersection is the minimum of corresponding counts.\n",
      " |      \n",
      " |      >>> c = Counter('abbb')\n",
      " |      >>> c &= Counter('bcc')\n",
      " |      >>> c\n",
      " |      Counter({'b': 1})\n",
      " |  \n",
      " |  __init__(self, iterable=None, /, **kwds)\n",
      " |      Create a new, empty Counter object.  And if given, count elements\n",
      " |      from an input iterable.  Or, initialize the count from another mapping\n",
      " |      of elements to their counts.\n",
      " |      \n",
      " |      >>> c = Counter()                           # a new, empty counter\n",
      " |      >>> c = Counter('gallahad')                 # a new counter from an iterable\n",
      " |      >>> c = Counter({'a': 4, 'b': 2})           # a new counter from a mapping\n",
      " |      >>> c = Counter(a=4, b=2)                   # a new counter from keyword args\n",
      " |  \n",
      " |  __ior__(self, other)\n",
      " |      Inplace union is the maximum of value from either counter.\n",
      " |      \n",
      " |      >>> c = Counter('abbb')\n",
      " |      >>> c |= Counter('bcc')\n",
      " |      >>> c\n",
      " |      Counter({'b': 3, 'c': 2, 'a': 1})\n",
      " |  \n",
      " |  __isub__(self, other)\n",
      " |      Inplace subtract counter, but keep only results with positive counts.\n",
      " |      \n",
      " |      >>> c = Counter('abbbc')\n",
      " |      >>> c -= Counter('bccd')\n",
      " |      >>> c\n",
      " |      Counter({'b': 2, 'a': 1})\n",
      " |  \n",
      " |  __missing__(self, key)\n",
      " |      The count of elements not in the Counter is zero.\n",
      " |  \n",
      " |  __neg__(self)\n",
      " |      Subtracts from an empty counter.  Strips positive and zero counts,\n",
      " |      and flips the sign on negative counts.\n",
      " |  \n",
      " |  __or__(self, other)\n",
      " |      Union is the maximum of value in either of the input counters.\n",
      " |      \n",
      " |      >>> Counter('abbb') | Counter('bcc')\n",
      " |      Counter({'b': 3, 'c': 2, 'a': 1})\n",
      " |  \n",
      " |  __pos__(self)\n",
      " |      Adds an empty counter, effectively stripping negative and zero counts\n",
      " |  \n",
      " |  __reduce__(self)\n",
      " |      Helper for pickle.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return repr(self).\n",
      " |  \n",
      " |  __sub__(self, other)\n",
      " |      Subtract count, but keep only results with positive counts.\n",
      " |      \n",
      " |      >>> Counter('abbbc') - Counter('bccd')\n",
      " |      Counter({'b': 2, 'a': 1})\n",
      " |  \n",
      " |  copy(self)\n",
      " |      Return a shallow copy.\n",
      " |  \n",
      " |  elements(self)\n",
      " |      Iterator over elements repeating each as many times as its count.\n",
      " |      \n",
      " |      >>> c = Counter('ABCABC')\n",
      " |      >>> sorted(c.elements())\n",
      " |      ['A', 'A', 'B', 'B', 'C', 'C']\n",
      " |      \n",
      " |      # Knuth's example for prime factors of 1836:  2**2 * 3**3 * 17**1\n",
      " |      >>> prime_factors = Counter({2: 2, 3: 3, 17: 1})\n",
      " |      >>> product = 1\n",
      " |      >>> for factor in prime_factors.elements():     # loop over factors\n",
      " |      ...     product *= factor                       # and multiply them\n",
      " |      >>> product\n",
      " |      1836\n",
      " |      \n",
      " |      Note, if an element's count has been set to zero or is a negative\n",
      " |      number, elements() will ignore it.\n",
      " |  \n",
      " |  most_common(self, n=None)\n",
      " |      List the n most common elements and their counts from the most\n",
      " |      common to the least.  If n is None, then list all element counts.\n",
      " |      \n",
      " |      >>> Counter('abracadabra').most_common(3)\n",
      " |      [('a', 5), ('b', 2), ('r', 2)]\n",
      " |  \n",
      " |  subtract(self, iterable=None, /, **kwds)\n",
      " |      Like dict.update() but subtracts counts instead of replacing them.\n",
      " |      Counts can be reduced below zero.  Both the inputs and outputs are\n",
      " |      allowed to contain zero and negative counts.\n",
      " |      \n",
      " |      Source can be an iterable, a dictionary, or another Counter instance.\n",
      " |      \n",
      " |      >>> c = Counter('which')\n",
      " |      >>> c.subtract('witch')             # subtract elements from another iterable\n",
      " |      >>> c.subtract(Counter('watch'))    # subtract elements from another counter\n",
      " |      >>> c['h']                          # 2 in which, minus 1 in witch, minus 1 in watch\n",
      " |      0\n",
      " |      >>> c['w']                          # 1 in which, minus 1 in witch, minus 1 in watch\n",
      " |      -1\n",
      " |  \n",
      " |  update(self, iterable=None, /, **kwds)\n",
      " |      Like dict.update() but add counts instead of replacing them.\n",
      " |      \n",
      " |      Source can be an iterable, a dictionary, or another Counter instance.\n",
      " |      \n",
      " |      >>> c = Counter('which')\n",
      " |      >>> c.update('witch')           # add elements from another iterable\n",
      " |      >>> d = Counter('watch')\n",
      " |      >>> c.update(d)                 # add elements from another counter\n",
      " |      >>> c['h']                      # four 'h' in which, witch, and watch\n",
      " |      4\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods defined here:\n",
      " |  \n",
      " |  fromkeys(iterable, v=None) from builtins.type\n",
      " |      Create a new dictionary with keys from iterable and values set to value.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from builtins.dict:\n",
      " |  \n",
      " |  __contains__(self, key, /)\n",
      " |      True if the dictionary has the specified key, else False.\n",
      " |  \n",
      " |  __eq__(self, value, /)\n",
      " |      Return self==value.\n",
      " |  \n",
      " |  __ge__(self, value, /)\n",
      " |      Return self>=value.\n",
      " |  \n",
      " |  __getattribute__(self, name, /)\n",
      " |      Return getattr(self, name).\n",
      " |  \n",
      " |  __getitem__(...)\n",
      " |      x.__getitem__(y) <==> x[y]\n",
      " |  \n",
      " |  __gt__(self, value, /)\n",
      " |      Return self>value.\n",
      " |  \n",
      " |  __iter__(self, /)\n",
      " |      Implement iter(self).\n",
      " |  \n",
      " |  __le__(self, value, /)\n",
      " |      Return self<=value.\n",
      " |  \n",
      " |  __len__(self, /)\n",
      " |      Return len(self).\n",
      " |  \n",
      " |  __lt__(self, value, /)\n",
      " |      Return self<value.\n",
      " |  \n",
      " |  __ne__(self, value, /)\n",
      " |      Return self!=value.\n",
      " |  \n",
      " |  __reversed__(self, /)\n",
      " |      Return a reverse iterator over the dict keys.\n",
      " |  \n",
      " |  __ror__(self, value, /)\n",
      " |      Return value|self.\n",
      " |  \n",
      " |  __setitem__(self, key, value, /)\n",
      " |      Set self[key] to value.\n",
      " |  \n",
      " |  __sizeof__(...)\n",
      " |      D.__sizeof__() -> size of D in memory, in bytes\n",
      " |  \n",
      " |  clear(...)\n",
      " |      D.clear() -> None.  Remove all items from D.\n",
      " |  \n",
      " |  get(self, key, default=None, /)\n",
      " |      Return the value for key if key is in the dictionary, else default.\n",
      " |  \n",
      " |  items(...)\n",
      " |      D.items() -> a set-like object providing a view on D's items\n",
      " |  \n",
      " |  keys(...)\n",
      " |      D.keys() -> a set-like object providing a view on D's keys\n",
      " |  \n",
      " |  pop(...)\n",
      " |      D.pop(k[,d]) -> v, remove specified key and return the corresponding value.\n",
      " |      \n",
      " |      If key is not found, default is returned if given, otherwise KeyError is raised\n",
      " |  \n",
      " |  popitem(self, /)\n",
      " |      Remove and return a (key, value) pair as a 2-tuple.\n",
      " |      \n",
      " |      Pairs are returned in LIFO (last-in, first-out) order.\n",
      " |      Raises KeyError if the dict is empty.\n",
      " |  \n",
      " |  setdefault(self, key, default=None, /)\n",
      " |      Insert key with a value of default if key is not in the dictionary.\n",
      " |      \n",
      " |      Return the value for key if key is in the dictionary, else default.\n",
      " |  \n",
      " |  values(...)\n",
      " |      D.values() -> an object providing a view on D's values\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from builtins.dict:\n",
      " |  \n",
      " |  __class_getitem__(...) from builtins.type\n",
      " |      See PEP 585\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods inherited from builtins.dict:\n",
      " |  \n",
      " |  __new__(*args, **kwargs) from builtins.type\n",
      " |      Create and return a new object.  See help(type) for accurate signature.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from builtins.dict:\n",
      " |  \n",
      " |  __hash__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30875"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c.most_common(50)[0][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "669"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "keywords = []\n",
    "counts = []\n",
    "top = c.most_common(669)\n",
    "for i in range(len(top)):\n",
    "    keywords.append(top[i][0])\n",
    "    counts.append(top[i][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = {}\n",
    "for i, word in enumerate(keywords):\n",
    "    d[word] = counts[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Reference proteome': 30875,\n",
       " 'Proteomics identification': 13651,\n",
       " 'Phosphoprotein': 10329,\n",
       " 'Membrane': 8060,\n",
       " 'Cytoplasm': 7829,\n",
       " 'Acetylation': 6250,\n",
       " 'Alternative splicing': 5855,\n",
       " '3D-structure': 5608,\n",
       " 'Nucleus': 5566,\n",
       " 'Repeat': 5035,\n",
       " 'Metal-binding': 4721,\n",
       " 'Transmembrane': 4447,\n",
       " 'Transmembrane helix': 4348,\n",
       " 'Nucleotide-binding': 3951,\n",
       " 'Direct protein sequencing': 3733,\n",
       " 'Coiled coil': 3654,\n",
       " 'Signal': 3622,\n",
       " 'Transport': 3379,\n",
       " 'Mitochondrion': 2963,\n",
       " 'ATP-binding': 2953,\n",
       " 'Hydrolase': 2912,\n",
       " 'Glycoprotein': 2890,\n",
       " 'Transferase': 2832,\n",
       " 'Disulfide bond': 2825,\n",
       " 'Ubl conjugation': 2801,\n",
       " 'Cell membrane': 2397,\n",
       " 'Zinc': 2223,\n",
       " 'Isopeptide bond': 2125,\n",
       " 'Disease variant': 1797,\n",
       " 'Cytoskeleton': 1787,\n",
       " 'Endoplasmic reticulum': 1737,\n",
       " 'RNA-binding': 1710,\n",
       " 'Secreted': 1630,\n",
       " 'Methylation': 1573,\n",
       " 'Transit peptide': 1471,\n",
       " 'Protein transport': 1443,\n",
       " 'Transcription': 1394,\n",
       " 'Calcium': 1333,\n",
       " 'Oxidoreductase': 1315,\n",
       " 'Transcription regulation': 1291,\n",
       " 'Zinc-finger': 1204,\n",
       " 'Kinase': 1152,\n",
       " 'Ribonucleoprotein': 1144,\n",
       " 'Cell projection': 1120,\n",
       " 'Golgi apparatus': 1118,\n",
       " 'Lipid metabolism': 1100,\n",
       " 'Magnesium': 1085,\n",
       " 'DNA-binding': 1029,\n",
       " 'Lipoprotein': 988,\n",
       " 'Mitochondrion inner membrane': 862}"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>keyword</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Reference proteome</td>\n",
       "      <td>30875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Proteomics identification</td>\n",
       "      <td>13651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Phosphoprotein</td>\n",
       "      <td>10329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Membrane</td>\n",
       "      <td>8060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>7829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>664</th>\n",
       "      <td>Pregnancy</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>Pontocerebellar hypoplasia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666</th>\n",
       "      <td>Thioether bond</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>667</th>\n",
       "      <td>Primary hypomagnesemia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>668</th>\n",
       "      <td>Bartter syndrome</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>669 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        keyword  count\n",
       "0            Reference proteome  30875\n",
       "1     Proteomics identification  13651\n",
       "2                Phosphoprotein  10329\n",
       "3                      Membrane   8060\n",
       "4                     Cytoplasm   7829\n",
       "..                          ...    ...\n",
       "664                   Pregnancy      1\n",
       "665  Pontocerebellar hypoplasia      1\n",
       "666              Thioether bond      1\n",
       "667      Primary hypomagnesemia      1\n",
       "668            Bartter syndrome      1\n",
       "\n",
       "[669 rows x 2 columns]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame({'keyword': keywords, 'count': counts})\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"data/keywords.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "79732fd3e39fcc86bf7c7aeb079145247bae47b6d1bafa505c129d7881fdf4a2"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
